### ISSUE TITLE
`@AnnotatedFor` in a stub file takes effect on a source class without `-AmergeStubsWithSource`

### ISSUE THREAD

**Issue by aosen-xiong** [eisop/checker-framework#1987]
**Reproducer** — compiled together with `-AuseConservativeDefaultsForUncheckedCode=source -Astubs=Lib.astub`, no `-AmergeStubsWithSource`:

```java
// Lib.java
public class Lib {
    public Object get() { return null; }
}
```

```java
// Lib.astub
@AnnotatedFor("nullness")
public class Lib {}
```

```java
// Use.java
@AnnotatedFor("nullness")
public class Use {
    void f(Lib lib) {
        @NonNull Object o = lib.get();
    }
}
```

**Expected:** the stub is ignored, so `Lib` is unchecked code. Its return gets conservative defaults and the caller is flagged:

```
Use.java:9: error: (assignment.type.incompatible)
```

**Actual:** the stub is applied, so `Lib` is treated as annotated for the Nullness Checker and gets type-checked:

```
Lib.java:5: error: (return.type.incompatible)
```

Adding a stub file silently opts a source file into checking, and errors are reported inside it.

**Cause:** `AnnotatedTypeFactory.getDeclAnnotations` adds stub-supplied declaration annotations unconditionally, whereas the type path applies the stub-precedence rule (`fromElement` consults `stubTypes` only when the element is not from source, or `-AmergeStubsWithSource` is passed).

**Comment by wmdietl** [comment]
Thanks for confirming! I already have a fix, I'll post it soon.

Options:
1. We document the current behavior, but the inconsistency bothers me.
2. We make declaration and type annotations consistent.
3. ?

I'll post an implementation of 2. soon.

(#1983 was an earlier version of this issue, with the same title?)

**Comment by aosen-xiong** [comment]
Sry, I shouldn't rename #1983. I was thinking about changing its content, but later realized a new issue might be a better idea. Title reverted.

I think we will need #1331's implementation at some point, as we will need to use `getDeclAnnotation` to determine whether the element is annotated.
